python/insights_ansible_playbook_signer/app.py

```diff
@@ -0,0 +1,179 @@
+        digest_file.write_bytes(play_digest)
+
+        result = crypto.sign_file(digest_file, key)
+        if not result.ok:
+            raise RuntimeError(f"Could not sign the digest: {result}")
+        return (temp_path / "digest.asc").read_bytes()
+
+
+def run() -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--version",
+        action="version",
+        version=get_version_from_package(),
+    )
+    parser.add_argument(
+        "--debug",
+        action="store_true",
+        help="Display logs",
+    )
+    keys = parser.add_mutually_exclusive_group(required=True)
+    keys.add_argument(
+        "--key",
+        type=pathlib.Path,
+        help="Path to private GPG key",
+    )
+    keys.add_argument(
+        "--remote-key",
+        type=str,
+        help="Name of a key for a remote signing server",
+    )
+    playbook = parser.add_mutually_exclusive_group(required=True)
+    playbook.add_argument(
+        "--playbook",
+        type=pathlib.Path,
+        help="Path to playbook",
+    )
+    playbook.add_argument(
+        "--stdin",
+        action="store_true",
+        help="Load playbook from stdin (the default)",
+    )
+    args = parser.parse_args()
+
+    # Load playbook to sign
+    raw_playbook: str = ""
+    if args.stdin:
+        with contextlib.suppress(KeyboardInterrupt):
+            raw_playbook = sys.stdin.read()
+    else:
+        raw_playbook = args.playbook.read_text()
+    if len(raw_playbook) == 0:
+        logger.error("Received empty playbook.")
+        raise RuntimeError("Received empty playbook.")
+
+    # Load plays
+    raw_plays: list[dict] = lib.parse_playbook(raw_playbook)
+    if not raw_plays:
+        raise lib.PreconditionError("Playbook contains no plays.")
+    logger.debug(f"Playbook contains {len(raw_plays)} plays.")
+
+    # Sign plays
+    plays: list[dict] = []
+    for i, raw_play in enumerate(raw_plays, 1):
+        play_name: str = raw_play.get("name", "???")
+        logger.debug(f"Preparing to sign play {play_name}.")
+        play: dict = copy.deepcopy(raw_play)
+
+        if "vars" not in play.keys():
+            logger.debug("Filling in missing 'vars' map.")
+            play["vars"] = {}
+        if "insights_signature_exclude" not in play["vars"].keys():
+            logger.debug("Filling in missing 'insights_signature_exclude' pair.")
+            play["vars"]["insights_signature_exclude"] = (
+                "/hosts,/vars/insights_signature"
+            )
+
+        if "insights_signature" not in play["vars"].keys():
+            # The 'clean_play' method requires this to be included.
+            # It will be overwritten later.
+            play["vars"]["insights_signature"] = ""
+
+        # Ensure 'tasks' are the last element
+        if "tasks" in play.keys():
+            play["tasks"] = play.pop("tasks")
+        else:
+            logger.warning("Play does not include 'tasks', is that correct?")
```

Comment:
I'd rather error out here, in case this situation is not wanted

Reply:
If we error out there, we won't be able to sign the revocation list YAML. If it's too messy for you, I'd rather create a follow-up card.

Reply:
IMHO the signer ought to work in an atomic way: either it signs both the playbook and the revocation list, or none of them.